File: firebase/functions/lib/events/live_meetings/breakouts/update_breakout_room_flag_status.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../../on_call_function.dart';
import '../../../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/firestore/event.dart';
import 'package:data_models/firestore/live_meeting.dart';
import 'package:data_models/firestore/membership.dart';
import 'package:data_models/utils.dart';

class UpdateBreakoutRoomFlagStatus
    extends OnCallMethod<UpdateBreakoutRoomFlagStatusRequest> {
  UpdateBreakoutRoomFlagStatus()
      : super(
          'UpdateBreakoutRoomFlagStatus',
          (json) => UpdateBreakoutRoomFlagStatusRequest.fromJson(json),
        );

  Future<void> _verifyCallerIsAuthorized(
    Event event,
    BreakoutRoom room,
    UpdateBreakoutRoomFlagStatusRequest request,
    CallableContext context,
  ) async {
    final communityMembershipDoc = await firestore
        .document(
          'memberships/${context.authUid}/community-membership/${event.communityId}',
        )
        .get();

    final membership = Membership.fromJson(
      firestoreUtils
          .fromFirestoreJson(communityMembershipDoc.data.toMap() ?? {}),
    );

    final isAuthorizedParticipant =
        room.participantIds.contains(context.authUid) &&
            request.flagStatus == BreakoutRoomFlagStatus.needsHelp;
    final isCreator = event.creatorId == context.authUid;
    final isAuthorized =
        membership.isMod || isCreator || isAuthorizedParticipant;
    if (!isAuthorized) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }
  }

  @override
  Future<String> action(
    UpdateBreakoutRoomFlagStatusRequest request,
    CallableContext context,
  ) async {
    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    await firestore.runTransaction((transaction) async {
      final breakoutRoomDoc = await firestore
          .document('${event.fullPath}/live-meetings/${event.id}'
              '/breakout-room-sessions/${request.breakoutSessionId}/breakout-rooms/${request.roomId}')
          .get();

      final requestedBreakoutRoom = BreakoutRoom.fromJson(
        firestoreUtils.fromFirestoreJson(breakoutRoomDoc.data.toMap() ?? {}),
      );

      await _verifyCallerIsAuthorized(
        event,
        requestedBreakoutRoom,
        request,
        context,
      );

      transaction.update(
        breakoutRoomDoc.reference,
        UpdateData.fromMap(
          jsonSubset(
            [BreakoutRoom.kFieldFlagStatus],
            firestoreUtils.toFirestoreJson(
              requestedBreakoutRoom
                  .copyWith(
                    flagStatus: request.flagStatus!,
                  )
                  .toJson(),
            ),
          ),
        ),
      );
    });

    return '';
  }
}

File: firebase/functions/src/events/live_meetings/breakouts/update_breakout_room_flag_status.ts
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import { OnCallMethod } from '../../../on_call_function';
import { firestore, firestoreUtils } from '../../../utils/infra/firestore_utils';
import { Event, Membership, BreakoutRoom, BreakoutRoomFlagStatus, membershipIsMod } from '../../../types';

interface UpdateBreakoutRoomFlagStatusRequest {
  eventPath: string;
  breakoutSessionId: string;
  roomId: string;
  flagStatus: BreakoutRoomFlagStatus;
}

export class UpdateBreakoutRoomFlagStatus extends OnCallMethod<UpdateBreakoutRoomFlagStatusRequest> {
  constructor() {
    super('UpdateBreakoutRoomFlagStatus', (json) => json as UpdateBreakoutRoomFlagStatusRequest);
  }

  async action(request: UpdateBreakoutRoomFlagStatusRequest, context: functions.https.CallableContext): Promise<string> {
    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });

    await firestore.runTransaction(async (transaction) => {
      const breakoutRoomDoc = await firestore
        .doc(
          `${(event as any).fullPath}/live-meetings/${(event as any).id}/breakout-room-sessions/${request.breakoutSessionId}/breakout-rooms/${request.roomId}`
        )
        .get();

      const breakoutRoom = firestoreUtils.fromFirestoreJson(breakoutRoomDoc.data() ?? {}) as unknown as BreakoutRoom;

      const membershipSnap = await firestore
        .doc(`memberships/${context.auth?.uid}/community-membership/${(event as any).communityId}`)
        .get();
      const membership = firestoreUtils.fromFirestoreJson(membershipSnap.data() ?? {}) as unknown as Membership;

      const isAuthorizedParticipant =
        ((breakoutRoom as any).participantIds ?? []).includes(context.auth?.uid) &&
        request.flagStatus === BreakoutRoomFlagStatus.needsHelp;
      const isCreator = (event as any).creatorId === context.auth?.uid;
      const isAuthorized = membershipIsMod((membership as any).status) || isCreator || isAuthorizedParticipant;

      if (!isAuthorized) {
        throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
      }

      transaction.update(breakoutRoomDoc.ref, { flagStatus: request.flagStatus });
    });

    return '';
  }
}
